pkg/controller/rollout/rollout_bluegreen.go

```diff
@@ -0,0 +1,487 @@
+		for example, if we had a rollout with 4 steps and CurrentStepIndex was 2
+		then, the user removed 3 steps from the plan, we can calculate NextStepIndex is 1 correctly,
+		but CurrentStepIndex remains 2, which could cause out of range.
+	*/
+	resetCurrentIndex := false
+	if int(bluegreenStatus.CurrentStepIndex) > len(c.Rollout.Spec.Strategy.BlueGreen.Steps) {
+		bluegreenStatus.CurrentStepIndex = nextIndex
+		resetCurrentIndex = true
+	}
+	currentStep := c.Rollout.Spec.Strategy.BlueGreen.Steps[bluegreenStatus.CurrentStepIndex-1]
+	if resetCurrentIndex || nextIndex != util.NextBatchIndex(c.Rollout, bluegreenStatus.CurrentStepIndex) && nextIndex > 0 {
+		currentIndexBackup := bluegreenStatus.CurrentStepIndex
+		currentStepStateBackup := bluegreenStatus.CurrentStepState
+		bluegreenStatus.CurrentStepIndex = nextIndex
+		bluegreenStatus.NextStepIndex = util.NextBatchIndex(c.Rollout, nextIndex)
+		nextStep := c.Rollout.Spec.Strategy.BlueGreen.Steps[nextIndex-1]
+		// if the Replicas between currentStep and nextStep is same, we can jump to
+		// the TrafficRouting step; otherwise, we should start from the Init step
+		if reflect.DeepEqual(nextStep.Replicas, currentStep.Replicas) && !resetCurrentIndex {
+			bluegreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			bluegreenStatus.CurrentStepState = v1beta1.CanaryStepStateTrafficRouting
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				bluegreenStatus.CurrentStepIndex, currentStepStateBackup, bluegreenStatus.CurrentStepState)
+		} else {
+			bluegreenStatus.LastUpdateTime = &metav1.Time{Time: time.Now()}
+			bluegreenStatus.CurrentStepState = v1beta1.CanaryStepStateInit
+			klog.Infof("rollout(%s/%s) step(%d) state from(%s) -> to(%s)", c.Rollout.Namespace, c.Rollout.Name,
+				bluegreenStatus.CurrentStepIndex, currentStepStateBackup, v1beta1.CanaryStepStateInit)
+		}
+		klog.Infof("rollout(%s/%s) canary step from(%d) -> to(%d)", c.Rollout.Namespace, c.Rollout.Name, currentIndexBackup, bluegreenStatus.CurrentStepIndex)
+		return true
+	}
+	return false
+}
+
+// cleanup after rollout is completed or finished
+func (m *blueGreenReleaseManager) doCanaryFinalising(c *RolloutContext) (bool, error) {
+	blueGreenStatus := c.NewStatus.BlueGreenStatus
+	if blueGreenStatus == nil {
+		return true, nil
+	}
+	// rollout progressing complete, remove rollout progressing annotation in workload
+	err := removeRolloutProgressingAnnotation(m.Client, c)
+	if err != nil {
+		return false, err
+	}
+
+	tr := newTrafficRoutingContext(c)
+	// execute steps based on the predefined order for each reason
+	nextStep := nextBlueGreenTask(c.FinalizeReason, blueGreenStatus.FinalisingStep)
+	// if current step is empty, set it with the first step
+	// if current step is end, we just return
+	if len(blueGreenStatus.FinalisingStep) == 0 {
+		blueGreenStatus.FinalisingStep = nextStep
```

Comment:
is it possible to move finalizing status to rollotu.status.conditions

Reply:
if the finished finalizing steps in saved in rollout.status.conditions, it is still possible to determine the current step in nextBlueGreenTask by iterating over taskSequence and find the last recorded step

Reply:
ok, i'll implement this logic in the coming patch